plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerPortExtensionTest.java

```diff
@@ -0,0 +1,284 @@
+    assertNotNull(portExtension.portLabel);
+    assertNotNull(portExtension.portText);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());
+  }
+
+  @Test
+  public void testDefaultPort() {
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowEmpty() {
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portText.getText().isEmpty());
+  }
+
+  @Test
+  public void testPortText_preventNonNumericCharacters() {
+    portExtension.portText.setText("80A80");
+    assertEquals("8080", portExtension.portText.getText());
+
+    portExtension.portText.setText("80 clipboard pasted 80");
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_numericCharacters() {
+    portExtension.portText.setText("808");
+    assertEquals("808", portExtension.portText.getText());
+
+    portExtension.portText.setText("8108");
+    assertEquals("8108", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowArbitrarilyLargeNumbers() {
+    portExtension.portText.setText("80808");
+    assertEquals("80808", portExtension.portText.getText());
+
+    portExtension.portText.setText("480808");
+    assertEquals("480808", portExtension.portText.getText());
+
+    portExtension.portText.setText("9480808");
+    assertEquals("9480808", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testAttributeSetOnValueChange() {
+    portExtension.portText.setText("12345");
+    assertEquals("12345", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 12345);
+
+    portExtension.portText.setText("54321");
+    assertEquals("54321", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 54321);
+
+    portExtension.portText.setText("0");
+    assertEquals("0", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 0);
+
+    portExtension.portText.setText("987654321");
+    assertEquals("987654321", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 987654321);
+  }
+
+  @Test
+  public void testAttributeDoesNotChangeOnBlockingNonNumericCharacters() {
+    portExtension.portText.setText("12345");
+    portExtension.portText.setText("12A345");
+    assertEquals("12345", portExtension.portText.getText());
+    verify(server, times(1)).setAttribute("appEngineDevServerPort", 12345);
+  }
+
+  @Test
+  public void testAttributeSetToZeroOnEmptyField() {
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portText.getText().isEmpty());
+    verify(server).setAttribute("appEngineDevServerPort", 0);
+  }
+
+  @Test
+  public void testNoPopUpDecorationForValidPorts() {
+    portExtension.portText.setVisible(true);
```

Comment:
Again, I'd prefer either looping over 0..65535 or use 1 instead of 1024 and 8080 (to test for extremal values).